### PR TITLE
Armour Penetration and Armour Damage Reduction nerf

### PR DESCRIPTION
**Pull request by RemieRichards**
- Armour damage reduction is now capped at 90% of the incoming damage, you will now, with the best armour setup in the game take 10% of the incoming damage.
- armour_penetration variable added to /item, this is counted for melee weapon and projectiles (AP ammo when?)
- Xenomorphs have a fixed 10 armour_penetration per melee attack
- The Ninja's Energy Katana has a 15 armour_penetration value (This both makes it stronger for them AND much better for any crew member who manages to turn it against them, as it will now pierce his suit too!)

Inspired by the Bay PR which uses horrible RNGesus to deal with armour penetration, since we use pure "80 armour = 80% damage reduction" maths, the armour penetration here does the same (80% armour - 20% armour_penetration = 60% armour)

### CODE REVIEW

**Comment by WJohn** [comment]
So armor penetration ignores however much percentage of armor someone has? Or is it a percentage of that percentage? If I were to hit someone with 90% melee resistance while I had 10 armor penetration, would they take damage as if they had 80% armor or 81%?

**Comment by RemieRichards** [comment]
@WJohn as if they had 80%. I did actually explain this in the op with an example but I'll repeat it.

80AA - 20 AP = 60A
A = Armour
AP = Armour Penetration

**Comment by Thunder12345** [comment]
Once this gets merged I might take a look at adding AP versions of the commonly used calibres to RnD. Maybe a couple of levels, for extra choice depending on whether mining's delivered.

**Comment by RemieRichards** [comment]
@Thunder12345 I'd appreciate that.
I didn't do it because I figured AP bullets should have unique sprites for their ammo, boxes, mags, etc.

**Comment by Iamgoofball** [comment]
fuck yes

**Comment by Ikarrus** [comment]
:+1: Especially for the armor cap

**Comment by Fox-McCloud** [comment]
But....but...my ACJ...

**Comment by RemieRichards** [comment]
@Fox-McCloud if you have an Administrative Jumpsuit you have access to actual godmode, surely? (as in the flag that prevents all damage)

**Comment by Fox-McCloud** [comment]
@RemieRichards 

Of course, I'm merely making a slight joking light of it--that said, it also prevents all stuns and other effects too; it's not big deal though, really--it's pure adminbus =p

**Comment by GunHog** [comment]
#9702 should be reverted before merging this, in my opinion.

**Comment by xxalpha** [comment]
:+1: for penetration

**Comment by Lumipharon** [comment]
Neat.

Also while on the whole I support the 90% reduction cap, not having that would allow neat situations where you could have special mobs/mechs/antags/whatever, that are only vulnerable to specialised weapons, because their resists are naturally 100% or even higher.

**Comment by WJohn** [comment]
You know, now that weapons can potentially have penetration properties, this makes a revert of kor's armor changes seem like an even better idea. Otherwise armor is basically going to be non-existent (it basically already is, but this just exacerbates the issue).

**Comment by Fox-McCloud** [comment]
@WJohn 

Yup, that could definitely be done, and I generally agree handling arpen on objects is probabbbly the better solution, long term--it's certainly more interesting.

File: code/modules/mob/living/living_defense.dm [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
+/mob/living/proc/run_armor_check(def_zone = null, attack_flag = "melee", absorb_text = null, soften_text = null, var/armour_penetration, var/penetrated_text)
 	var/armor = getarmor(def_zone, attack_flag)
+
+	//the if "armor" check is because this is used for everything on /living, including humans
+	if(armor && armour_penetration)
+		armor = max(0, armor - armour_penetration)
+		if(penetrated_text)
+			src << "<span class='userdanger'>[penetrated_text]</span>"
+		else
+			src << "<span class='userdanger'>Your armor was penetrated!</span>"
```
Comment (phil235) [inline code/modules/mob/living/living_defense.dm]: Do we really need a default message for that? With this added, someone being attacked will receive three different messages per hit taken, that's a bit spammy.

Comment (RemieRichards) [inline code/modules/mob/living/living_defense.dm]: I feel being told your armour just failed on you is an important thing to be told.
plus the others have default texts so, for consistency/overriding.

**Comment by WJohn** [comment]
I've been wanting stuff like different bullet calibers for a long time. Imagine armor piercing .357 rounds that do less damage by themselves (45 or 50 brute) but have huge armor piercing qualities (while also being more expensive). Or perhaps anti-personnel rounds that are MORE damaging (75-80 brute) but could have negative armor piercing properties, making them significantly worse against armored people.

Heck, apply this to all bullet types (10mm for the pistol, 45 for the c20r and so on), let us mix and match what goes in ammo boxes and in magazines and we'd have a pretty neat system. It'd definitely be interesting to watch the syndies invest in choosing anti-personnel and then hit medbay, or choose armor piercing and hitting security, depending on their guns.

**Comment by Fox-McCloud** [comment]
@WJohn 

Your dream has come true then; Remie's system allows for that very thing.

**Comment by freerealestate** [comment]
Why the cap? Are any armour values actually at 100? If they are, wouldn't it be a better idea to change them instead?

**Comment by RemieRichards** [comment]
@freerealestate it's not that any armours have 100% protection, it's that they stack
10% sec jumpsuit + fully upgraded mining rig is about 80-90% iirc, so all it takes is another armour or another combo being added (without any intention of them stacking) and you end up with invulnerability.

plus invulnerability should not be obtainable, it's a game.

**Comment by freerealestate** [comment]
>it's that they stack
>10% sec jumpsuit + fully upgraded mining rig is about 80-90% iirc, so all it takes is another armour or another combo being added (without any intention of them stacking) and you end up with invulnerability.

Out of curiosity, does armour only stack on the body parts covered, or the entire body?

>plus invulnerability should not be obtainable, it's a game.

I agree, though I can't see a scenario where you could lose with 90% damage reduction

**Comment by RemieRichards** [comment]
@freerealestate if the proc called for the damage check is specified a limb, as in you are targetting and hitting only, an arm lets say, then it only checks armours that say they cover arms.
if it is not specified a limb, it checks every piece of clothing the user is wearing, and produces and average amount of "armour" (say 80, as it is always interpreted as a percentage) for that attack to be reduced by

**Comment by WJohn** [comment]
With kor's current armor nerf it's no longer possible to actually reach those numbers, but it was in the past. I think you were capable originally of flipping a security hood up even while wearing a sec helmet, and that brought your head's melee resistance to 100%.

**Comment by Jordie0608** [comment]
Changelog

**Comment by Steelpoint** [comment]
Either way this is good, but we will need to look over armour values considering now that most armour defences are too low to really make a difference with any armour penetration weapons.

**Comment by optimumtact** [comment]
If you're a security player you can't comment okay

**Comment by Miauw62** [comment]
im very happy with the 90% armor cap and rather confused as to why nobody thought of implementing it earlier.

**Comment by Steelpoint** [comment]
It had been suggest for quite some time as a response to the fact that multiple in game methods can get you to that 100 number. Ultimantly it just took until now for this suggestion to be implemented.

**Comment by RemieRichards** [comment]
tbh I saw bay's AP pr, hated it because of rngesus, decided to make one for us (obvi better :P) and then noticed the ```if(armour >= 100)``` return 0 in human ```attacked_by``` and it reminded me, so I just swapped those 2 lines out (it's in species aswell) with a ```armor = min(90, armor)```

But yes, the intention (and reason why very few things in this PR atm actually use the AP system) is to get armour values put back up (Maybe not the original values, but definitely not the current ones)

AP also allows us to have weapons that are strong against armour without being Death's Scythe to people without armour.

**Comment by Steelpoint** [comment]
The only concern I have is that any armour value revert might be made meaningless if the armour piercing changes to antag weaponry makes the revert useless. You go from a, for example, three hit crit with weapon X, but you rebuff the armour but add in a armour piercing value to weapon X that still results in a three hit crit.

**Comment by RemieRichards** [comment]
@Steelpoint my intention is for AP to be rare (which is why in this PR it's only been added to two rare powerful antags (and in the case of one of these antags, can actually be a negative for them))

**Comment by Durkel** [comment]
I think being augmented by robotics reduces damage by a flat number, five or something. Might want to look at that while you're at it.

**Comment by KorPhaeron** [comment]
@remierichards

Jumpsuits stack with armor? I thought it just picked the higher value

**Comment by Steelpoint** [comment]
Sec Jumpsuits have 10 melee protection which is added to any armour that is being worn.

**Comment by RemieRichards** [comment]
@KorPhaeron 

getarmor() checks all clothing you're wearing. if you give getarmor an organ or zone to check for, it will only check clothing with the right flags to be covering that body part, if you don't give it an organ argument it just averages all the armour you're wearing.

tl;dr yes it stacks.

**Comment by Steelpoint** [comment]
Wait a second, I don't recall the Ninja's eblade being something that can be dropped. Unless that was changed since I never see Ninja in game.

**Comment by freerealestate** [comment]
Yeah, that's what I thought too, Steelpoint

**Comment by Lumipharon** [comment]
ninja got completely overhauled by remie. Check it out.

**Pull request merged**